### PR TITLE
DWR-1009 Redshift research - just to keep track of my changes

### PR DESCRIPTION
**Pull request by agorina**
@malaffoon and @jtreuting, this is my way to provide an update on the progress.
Please review and let me know if you have any comments.

### CODE REVIEW

File: aggregate_reporting/PerformanceTuningTripsAndTricks.md [hunk 1]
```diff
@@ -0,0 +1,168 @@
+You can find additional resources here:
```
Comment (agorina) [inline aggregate_reporting/PerformanceTuningTripsAndTricks.md]: I have decided to start this document to keep track of what I have been doing.

File: aggregate_reporting/PerformanceTuningTripsAndTricks.md [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+- [AWS Big Data Blog, Amazon Redshift Engineering’s Advanced Table Design Playbook](https://aws.amazon.com/blogs/big-data/amazon-redshift-engineerings-advanced-table-design-playbook-preamble-prerequisites-and-prioritization/)
+
+## Prerequisites
+Admin queries refer to tables by their object ID (OID). You can get this OID in multiple ways, below is one of them:
+```sql
+SELECT oid, relname FROM pg_class WHERE relname='your_table_name_here';
+```
+
+### Admin schema
+Create admin schema in the Amazon Redshift cluster database you’re optimizing.
```
Comment (agorina) [inline aggregate_reporting/PerformanceTuningTripsAndTricks.md]: I do not know why this is not provided out of the box, but it is not.
And there is no single place to have all these views, it came from different sources.

File: aggregate_reporting/PerformanceTuningTripsAndTricks.md [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+```sql
+create view admin.WLM_QUERY_STATE_VW as
+select query, (service_class-5) as queue, slot_count, trim(wlm_start_time) as start_time, trim(state) as state, trim(queue_time) as queue_time, trim(exec_time) as exec_time
+from stv_wlm_query_state;
+```
+
+### Troubleshooting by scenarios
+
+##### Scenario: “There are no specific reports of slowness, but I want to ensure I’m getting the most out of my cluster by performing a review on all tables.”
+Returns table information for all scanned tables
```
Comment (agorina) [inline aggregate_reporting/PerformanceTuningTripsAndTricks.md]: I am planning on expanding these use cases.

File: aggregate_reporting/PerformanceTuningTripsAndTricks.md [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+  trim(split_part(l.event, ':', 1))                                                                                                                               AS event,
+  substring(trim(l.solution), 1, 60)                                                                                                                              AS solution,
+  max(l.query)                                                                                                                                                    AS sample_query,
+  count(DISTINCT l.query)
+FROM stl_alert_event_log AS l
+  LEFT JOIN stl_scan AS s ON s.query = l.query AND s.slice = l.slice AND s.segment = l.segment
+  LEFT JOIN stl_dist AS d ON d.query = l.query AND d.slice = l.slice AND d.segment = l.segment
+  LEFT JOIN stl_bcast AS b ON b.query = l.query AND b.slice = l.slice AND b.segment = l.segment
+WHERE l.userid > 1
+      AND l.event_time >= dateadd(h, -1, localtimestamp)
```
Comment (agorina) [inline aggregate_reporting/PerformanceTuningTripsAndTricks.md]: @malaffoon , thank you for helping with the timestamp thing here.

File: aggregate_reporting/sql/V1_0_0_0__ddl.sql [hunk 5]
```diff
@@ -189,21 +189,21 @@ CREATE TABLE  district (
 ) DISTSTYLE ALL;
 
 CREATE TABLE school (
-  id integer encode runlength PRIMARY KEY SORTKEY,
+  id integer encode raw PRIMARY KEY SORTKEY,
```
Comment (agorina) [inline aggregate_reporting/sql/V1_0_0_0__ddl.sql]: It is my understanding that the sort and distribution keys should not be encoded.

File: aggregate_reporting/sql/V1_0_0_0__ddl.sql [hunk 6]
```diff
@@ -220,23 +220,24 @@ CREATE TABLE ethnicity (
...
   id bigint encode delta PRIMARY KEY,
-  school_id integer encode runlength NOT NULL,
-  student_id bigint encode delta32k NOT NULL DISTKEY,
-  asmt_id bigint encode lzo NOT NULL,
+  school_id integer encode raw  NOT NULL,
+  student_id bigint encode raw  NOT NULL DISTKEY,
+  asmt_id bigint encode raw  NOT NULL,
   grade_id smallint encode lzo NOT NULL,
-  school_year smallint encode lzo NOT NULL,
+  asmt_grade_id smallint encode lzo NOT NULL,
```
Comment (agorina) [inline aggregate_reporting/sql/V1_0_0_0__ddl.sql]: I added it back here, I think it makes it easier for Redshift. Still testing.

File: aggregate_reporting/sql/_sample_aggreage_reports.sql [hunk 7]
```diff
@@ -1,78 +1,257 @@
 ------------------------------------------------------------------------------------------------------------
 -- Achievement level report grades 3 and 4, years 2017 and 2018 by schools and withing each school by asmt grade and gender
 ------------------------------------------------------------------------------------------------------------
+-- State by gender
```
Comment (agorina) [inline aggregate_reporting/sql/_sample_aggreage_reports.sql]: This is my attempt to add more focused (narrow view on the data) queries.
I am planning on using them all in the jmeter and replace hardcoded values with the parameters from a data bank.

File: aggregate_reporting/sql/_redshift_load_from_staging.sql [hunk 8]
```diff
@@ -97,6 +97,7 @@ INSERT INTO fact_student_ica_exam (
   student_id,
   asmt_id,
   grade_id,
+  asmt_grade_id,
```
Comment (malaffoon) [inline aggregate_reporting/sql/_redshift_load_from_staging.sql]: denormalizing to improve perf?

Comment (agorina) [reply aggregate_reporting/sql/_redshift_load_from_staging.sql]: That was an idea. Still not sure if it is needed.

**Pull request merged**